### PR TITLE
os/bluestore: do not update used_blocks with bluefs_extents if bluefs is not available

### PR DESCRIPTION
**Pull request by hualongfeng**
When I disabled bluefs, the bluefs point is nullptr, but is still used. So I add a judge for it and it works well.
It caused the problem:
*** Caught signal (Segmentation fault) **
 in thread 7f733db01d80 thread_name:ceph-osd
 ceph version 16.0.0-4201-g2ccd47711b (2ccd47711b73b09069e6964730532fc3cf6e7382) pacific (dev)
 1: (()+0x153c0) [0x7f733e1a73c0]
 2: (pthread_mutex_lock()+0x4) [0x7f733e19dfc4]
 3: (BlueFS::get_block_extents(unsigned int, interval_set<unsigned long, std::map>*)+0x40) [0x562384a57680]
 4: (BlueStore::_fsck_on_open(BlueStore::FSCKDepth, bool)+0x743) [0x5623849b1143]
 5: (BlueStore::_fsck(BlueStore::FSCKDepth, bool)+0x32a) [0x5623849be1da]
 6: (BlueStore::mkfs()+0xc77) [0x5623849c0967]
 7: (OSD::mkfs(ceph::common::CephContext*, ObjectStore*, uuid_d, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)+0x1b2) [0x5623843e85c2]
 8: (main()+0x1563) [0x56238436b3f3]
 9: (__libc_start_main()+0xf3) [0x7f733dc0e0b3]
 10: (_start()+0x2e) [0x5623843b734e]
2020-08-10T00:54:09.896+0000 7f733db01d80 -1 *** Caught signal (Segmentation fault) **
 in thread 7f733db01d80 thread_name:ceph-osd

 ceph version 16.0.0-4201-g2ccd47711b (2ccd47711b73b09069e6964730532fc3cf6e7382) pacific (dev)

Signed-off-by: Feng Hualong <hualong.feng@intel.com>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -8014,8 +8014,10 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
   _fsck_collections(&errors);
   used_blocks.resize(fm->get_alloc_units());
 
-  int r = bluefs->get_block_extents(bluefs_layout.shared_bdev, &bluefs_extents);
-  ceph_assert(r == 0);
+  if(bluefs) {
+    int r = bluefs->get_block_extents(bluefs_layout.shared_bdev, &bluefs_extents);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:8018]: Please add space between "if" and "(".

**Comment by hualongfeng** [comment]
Jenkins, test this please

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -8014,8 +8014,10 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
   _fsck_collections(&errors);
   used_blocks.resize(fm->get_alloc_units());
 
-  int r = bluefs->get_block_extents(bluefs_layout.shared_bdev, &bluefs_extents);
-  ceph_assert(r == 0);
+  if (bluefs) {
+    int r = bluefs->get_block_extents(bluefs_layout.shared_bdev, &bluefs_extents);
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc:8018]: @hualongfeng could you also move `interval_set<uint64_t> bluefs_extents;` and the `for` iterating over `bluefs_extents`  into this `if` block as well?

also, i think we could be more specific in the title:

>  fixed a bug when unusing bluefs

how about:

> do not update used_blocks with bluefs_extents if bluestore is not available

**Comment by hualongfeng** [comment]
@tchaikov That's a good idea. I changed it and used that title. But I replace the "bluestore is not available" as "bluefs is not available".

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -8006,24 +8006,27 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
...
+    ceph_assert(r == 0);
+    for (interval_set<uint64_t>::iterator it = bluefs_extents.begin(); it != bluefs_extents.end(); ++it) {
+
+      apply_for_bitset_range(it.get_start(), it.get_len(), alloc_size, used_blocks,
+        [&](uint64_t pos, mempool_dynamic_bitset& bs) {
+          ceph_assert(pos < bs.size());
+          bs.set(pos);
+        }
+      );
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc:8028]:
```suggestion
    for (auto [start, len] : bluefs_extents) {
      apply_for_bitset_range(start, len, alloc_size, used_blocks,
        [&](uint64_t pos, mempool_dynamic_bitset& bs) {
          ceph_assert(pos < bs.size());
          bs.set(pos);
        }
      );
```

could you take this opportunity to refactor this `for` loop as well?

**Comment by hualongfeng** [comment]
@tchaikov I have changed the for loop.

**Comment by hualongfeng** [comment]
Jenkins, test this please

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (hualongfeng) [reply src/os/bluestore/BlueStore.cc:8019]: @aclamk I have added space between "if" and "(". (I remember that I reply the change but it seems not  success)

**Comment by tchaikov** [comment]
@aclamk could you take another look?

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-08-18_13:00:33-rados-wip-kefu-testing-2020-08-18-1658-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-19_06:47:30-rados-wip-kefu-testing-2020-08-19-1141-distro-basic-smithi/

**Comment by hualongfeng** [comment]
@tchaikov I test [the failed test case](http://qa-proxy.ceph.com/teuthology/kchai-2020-08-19_06:47:30-rados-wip-kefu-testing-2020-08-19-1141-distro-basic-smithi/5359035/teuthology.log) on my teuthology environment, but it passed. I cannot reproduce that problem.

**Comment by tchaikov** [comment]
> @tchaikov I test [the failed test case](http://qa-proxy.ceph.com/teuthology/kchai-2020-08-19_06:47:30-rados-wip-kefu-testing-2020-08-19-1141-distro-basic-smithi/5359035/teuthology.log) on my teuthology environment, but it passed. I cannot reproduce that problem.

@hualongfeng thanks for looking into the failed tests. neither do i think they are related. i just wanted to wait for Adam's final review.

**Pull request merged**